universum/modules/code_report_collector.py

```diff
@@ -14,6 +15,12 @@
 from .structure_handler import HasStructure
 
 
+class ProjectConfig(TypedDict, total=False):  # TODO: define a proper class for this
```

Comment:
ProjectConfig seems like a complicated data structure, where value types for certain keys are important
consider refactoring as a full-fledged class

Reply:
project_configs is just Variations merged into a flat list. Hence, ProjectConfig is just one configuration.
Its keys are described here: https://universum.readthedocs.io/en/latest/configuring.html#common-variations-keys

Reply:
how about we define all of the possible keys in ProjectConfig TypedDict?
Would that be useful?

Reply:
Yes, for sure.
Can we somehow connect it to documentation?
For example, is there possibility to annotate and document it within the code?

Reply:
The only thing, this is not the place. This module is just a sidekick of a Launcher.

Reply:
> Can we somehow connect it to documentation?
> For example, is there possibility to annotate and document it within the code?
I need to investigate that. This ProjectConfig(TypedDict) mypy class doesn't exist in runtime (it's just dict), but we might still put docstrings there.

Reply:
> The only thing, this is not the place. This module is just a sidekick of the Launcher.

I was considering putting the new class to Variations, but that one seems more generic. Launcher is a fine location too, I guess.

Reply:
Should I do this as part of current PR?

Reply:
> I was considering putting the new class to Variations, but that one seems more generic. Launcher is a fine location too, I guess.

Variations is more user friendly. Launcher is the one that mostly uses those keys.
I think we should be user friendly.

I agree that Variations is generic, but, frankly, no one will ever use it for anything except creating configs for Launcher.
Also, I don't think we will ever see anything, but Variations, passed as configs to Launcher.

Reply:
> Should I do this as part of current PR?

I think it deserves separate commit.

Reply:
I think so too, especially if I am going to narrow down Variations (which seems like a good idea to me now, but we'll see)

Reply:
> I need to investigate that. This ProjectConfig(TypedDict) mypy class doesn't exist in runtime (it's just dict), but we might still put docstrings there.

Tested and confirmed that sphinx works even for helper classes like this.